### PR TITLE
Allow environment variable to indicate we can load unsigned kexts.

### PR DESCRIPTION
**Pull request by dfjdejulio**
Brew formulae that require unsigned kexts currently refuse to work on Yosemite because of the default of requiring signed kexts. However, if you boot Yosemite with "kext-dev-mode=1" among the boot arguments, unsigned kexts work just fine, and you can make this persistent for a given machine via the "nvram" command.

This patch lets you set the HOMEBREW_KEXT_DEV_MODE environment variable to indicate that you've configured your system to permit unsigned kexts, satisfying the requirement.

### CODE REVIEW

**Comment by DomT4** [comment]
To be a little blunt, and I'll preface this by saying my opinion counts for nothing here and the maintainers are free to ignore the hoot out of me, but I *really* don't want to do this, or support this, or even touch this.

I consider the unsigned kext ban to be a solid piece of security, and frankly, something developers had *two years* notice to do something about before Apple implemented it strictly. The console warnings were in the very first few Mavericks Developer Previews, and the strict ban was in the very first few Yosemite Developer Previews. It's not like this change snuck up on anyone. Yes, the system is frustrating and it isn't the easiest process to obtain a kext signing cert in the world, but I think on balance not having anyone being able to willy-nilly insert kernel extensions into a user's system unguarded isn't an awful measure.

Specifically here, I think this block:

> Loading unsigned kexts
      requires booting with the "kext-dev-mode=1" argument, which can
      be placed in the "boot-args" nvram variable.  To indicate that
      you're running with this configuration and can load unsigned kexts,
      set the environment variable "HOMEBREW_KEXT_DEV_MODE".

Will lead to people just going *"Oh sure, I'll just bash that command into my Terminal, reboot and boom, I can run unsigned kexts again, yay"* because the vast majority of people won't stop to think *Oh, this means anyone can run any kext on my system with no safeguards in place*. People love the easy option, and that's no diss on human nature, it's just how we are.

I also don't think the gain here outweighs the risks. If we look at the formulae currently containing a ` UnsignedKextRequirement `:

* osxfuse
* tuntap
* horndis
* fuse4x

* Osxfuse currently has a signed, available, self-updating upstream binary which you can install via the website or with ` brew cask install osxfuse `.

* Tuntap currently has a signed, available upstream binary which you can get from Sourceforge or via ` brew cask install tuntap `.

* Horndis is not code-signed at this point. I don't know if usage is really wide enough to justify changing the policy so radically:

```
Code Signing Failure: not code signed
HoRNDIS.kext appears to be loadable (not including linkage for on-disk libraries).
ERROR: invalid signature for com.joshuawise.kexts.HoRNDIS, will not load
```

* fuse4x is dead for all intents and purposes in terms of upstream support and is extremely slowly merging into osxfuse. Fuse4x themselves have been saying [don't use us, use osxfuse](https://groups.google.com/forum/#!topic/fuse4x/Pi55U_K4HgU) for ages. I'd consider a better course of action to be pointing this out to upstream projects that are *still* using fuse4x.

Honestly, and again, my own views here - When I initially discussed and later wrote the requirement I more envisioned a place where when 10.11 lands, whatever Apple decide to call it, Homebrew can stop shipping formulae that use kexts at all and leave it to the Cask. The Cask is brilliantly suited to distributing kexts. It does a considerably better job of cleaning up system remnants (i.e. stuff that lives outside its prefix) than Homebrew/Homebrew does, and it happily ships binary packages.

I don't think easier access to running unsigned kexts is a gain, I actually consider it a negative, and whilst I appreciate the effort and sentiment behind this PR, so please don't take this as a personal thing, I just think the root outcome is a negative one on this singular issue.

*tl;dr - Dom thinks this is a terrible idea. Dom's opinions have no bearing on the maintainers' decisions.*

**Comment by dfjdejulio** [comment]
I'll explain why I bothered to do this.

In general, I prefer to run code I compile myself instead of installing binaries someone else has built, so I do not like casks and the like if I can reasonably avoid them.  So I always set the environment variable that tells Homebrew "build everything from source", and I'd very much like to not make an exception for kexts.

One reason for that is, on some of my systems, I'm running prerelease versions of OS X.  The easier it is for me to build open source projects against the prerelease libraries and kernels with the prerelease dev tools, the more likely I'll catch an issue before the release, and may be able to contribute upstream fixes before most users are impacted.

If someone wants to add some verbiage to the effect of "don't do this unless you're a crazy person and understand the risks", I'd have no objection to that.

But ultimately, my *understanding* is that the requirement logic is supposed to just indicate whether a requirement, of being able to install unsigned kexts in this case, can technically be met, and I corrected what I thought was a simple error in that logic -- the ban on unsigned kexts is *not* actually a strict one, and a Yosemite system *can* be configured to permit them.

**Comment by mikemcquaid** [comment]
> In general, I prefer to run code I compile myself instead of installing binaries someone else has built, so I do not like casks and the like if I can reasonably avoid them. So I always set the environment variable that tells Homebrew "build everything from source", and I'd very much like to not make an exception for kexts.

Pedantry but you're not really building from source yourself if you're doing it through Homebrew. If you're comfortable fixing the issues that crop up that way though, feel free.

> But ultimately, my understanding is that the requirement logic is supposed to just indicate whether a requirement, of being able to install unsigned kexts in this case, can technically be met, and I corrected what I thought was a simple error in that logic -- the ban on unsigned kexts is not actually a strict one, and a Yosemite system can be configured to permit them.

Only by enabling an intentionally added security feature. For that reason I'm :-1: on pointing users to circumvention measures where there's decent upstream binaries available.

**Comment by DomT4** [comment]
> In general, I prefer to run code I compile myself instead of installing binaries someone else has built, so I do not like casks and the like if I can reasonably avoid them.

The Cask isn't all that much different to Homebrew/Homebrew in that sense. The Cask distributes binaries produced by the app/tool/utility/etc producer; Homebrew distributes binaries its own bot generates. When things are poured from a bottle that's just a pre-generated binary.

This point actually reminds me of something I said to Mike, one of the maintainers a while back, when talking about the history of Homebrew, I mentioned kinda hating trusting upstream binaries and pondered whether that mistrust of binaries was part of why Homebrew was created.

FWIW; The Homebrew binaries are relatively simple tarballs. They're autogenerated based on the formula preferences specified, prefix=blah, and so on. The build process and environmental variables used are all in the test-bot code, and formulae code. Homebrew's generated binaries are about as transparent as you can get; although I fully appreciate and understand why people can and may wish to choose to opt out of those.

>  the more likely I'll catch an issue before the release, and may be able to contribute upstream fixes before most users are impacted.

This kinda makes you different from the typical examples of Bob, Sally, Dave, Generic Name here, etc. You're running both pre-release, and you contribute fixes to upstreams where possible. That development background is *why* the nvram argument does exist - For developers. It was, as far as Apple have said, never intended as an easy way for normal users to circumvent the kext ban.

> But ultimately, my understanding is that the requirement logic is supposed to just indicate whether a requirement

To be honest, the requirements kinda bend to fit whatever is required of them. They're incredibly flexible. But AFAIK, from when it was written, I don't think the intention was supposed for it to be ambiguous. It is designed to stop people circumventing Apple's strict kext ban *(The "strict" language comes from one of their beta release notes as far as I remember)*.

If people want to go around it, we can't stop them just plonking the words "Yosemite Kext Ban" into their favourite search engine and then removing the requirement from the affected formulae, but I largely tend to think circumventing Apple's upstream decisions around the kernel is asking for future maintenance headaches for users and the maintainers here. Again, all my own opinions on this.

If developers want to disable it, and are comfortable with the ramifications of that, that's cool, nothing stopping them from doing so. But I'm not personally completely comfortable telling people how to avoid it every time they hit the requirement message, because even if we inserted a *"Don't do this please unless you know what you're doing!"* message into the requirement pretty quickly people would ignore the warning and follow the remaining instructions regardless.

**Comment by dfjdejulio** [comment]
> The Cask distributes binaries produced by the app/tool/utility/etc producer; Homebrew distributes
> binaries its own bot generates.

Not if you use the command line switch or set the environment variable to always build from source, as I do.

**Comment by dfjdejulio** [comment]
Hm.  I think I _would_ suggest not pre-building the binaries for unsigned kexts, so they're only there for people who do the building themselves.

If folks don't like the idea of documenting how to do this, maybe the requirement should check the boot arguments directly?  Then it would "just work" on any system that was _already_ configured to load them.  I think that would be a little less friendly to people who legitimately need this, but it would be better at discouraging the rank and file.

**Comment by mikemcquaid** [comment]
> If folks don't like the idea of documenting how to do this, maybe the requirement should check the boot arguments directly? Then it would "just work" on any system that was already configured to load them. I think that would be a little less friendly to people who legitimately need this, but it would be better at discouraging the rank and file.

I'd rather these people used a custom fork or tap to install these formulae.

**Comment by xu-cheng** [comment]
@dfjdejulio since this feature is very custom to yourself. I recommend you create [a custom external homebrew command](https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/External-Commands.md). So you can overwrite homebrew security failsafe and then call brew's install method to install whatever unsafe prerelease formulae as you see fit.

**Comment by dfjdejulio** [comment]
Well, it wasn't *intended* to be custom to myself -- it's just a way to indicate that the "unsigned-kexts" requirement that some packages have can actually be met even if you're on Yosemite -- but I'll look into that custom command documentation and I thank you for the reference.

**Comment by xu-cheng** [comment]
Since it's a security manner. So such feature should never be implemented in the core by default. As for users who need such feature and probably be kernel developers should be aware of kext-dev-mode=1 themselves and be able to do any needed custom by their own. 

As for external command, [here](https://github.com/neovim/homebrew-neovim/blob/master/bin/brew-head-audit.rb) is a example I used to allow homebrew audit head-only formula. Hope this can be helpful to you.

**Comment by mikemcquaid** [comment]
Closing this as it won't be included as-is. Sorry!

**Comment by DomT4** [comment]
> Closing this

Tiny heads up, you didn't close this :wink:
